src/Utils.cpp

```diff
@@ -468,11 +470,95 @@ utils::escapeBlacklistedHtml(const QString &rawStr)
+utils::markdownToHtml(const QString &text, bool rainbowify)
+{
+        const auto str = text.toUtf8();
+        cmark_node *const node =
+          cmark_parse_document(str.constData(), str.size(), CMARK_OPT_UNSAFE);
+
+        if (rainbowify) {
+                // create iterator over node
+                cmark_iter *iter = cmark_iter_new(node);
+
+                cmark_event_type ev_type;
+
+                // First loop to get total text length
+                int textLen = 0;
+                while ((ev_type = cmark_iter_next(iter)) != CMARK_EVENT_DONE) {
+                        cmark_node *cur = cmark_iter_get_node(iter);
+                        // only text nodes (no code or semilar)
+                        if (cmark_node_get_type(cur) != CMARK_NODE_TEXT)
+                                continue;
+                        // count up by length of current node's text
+                        QTextBoundaryFinder tbf(QTextBoundaryFinder::BoundaryType::Grapheme,
+                                                QString(cmark_node_get_literal(cur)));
+                        while (tbf.toNextBoundary() != -1)
+                                textLen++;
+                }
+
+                // create new iter to start over
+                cmark_iter_free(iter);
+                iter = cmark_iter_new(node);
+
+                // Second loop to rainbowify
+                int charIdx = 0;
+                while ((ev_type = cmark_iter_next(iter)) != CMARK_EVENT_DONE) {
+                        cmark_node *cur = cmark_iter_get_node(iter);
+                        // only text nodes (no code or semilar)
+                        if (cmark_node_get_type(cur) != CMARK_NODE_TEXT)
+                                continue;
+
+                        // get text in current node
+                        const char *tmp_buf = cmark_node_get_literal(cur);
+                        std::string nodeText(tmp_buf);
+                        auto qNodeText = QString::fromStdString(nodeText);
+                        // create buffer to append rainbow text to
+                        std::string buf;
+                        int boundaryStart = 0;
+                        int boundaryEnd   = 0;
+                        // use QTextBoundaryFinder to iterate ofer graphemes
+                        QTextBoundaryFinder tbf(QTextBoundaryFinder::BoundaryType::Grapheme,
+                                                qNodeText);
+                        while ((boundaryEnd = tbf.toNextBoundary()) != -1) {
+                                // Split text to get current char
+                                auto curChar =
+                                  qNodeText.mid(boundaryStart, boundaryEnd - boundaryStart);
+                                boundaryStart = boundaryEnd;
+                                // Don't rainbowify spaces
+                                if (curChar == " ") {
```

Comment:
I'd do curChar.trimmed().isEmpty(), because that checks for all whitespace chars. And then in the push back append the curChar